internal/newstate/graph.go

```diff
@@ -0,0 +1,210 @@
+package newstate
+
+import (
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+// listener represents a listener of the Gateway resource.
+// FIXME(pleshakov) For now, we only support HTTP listeners.
+type listener struct {
+	// Source holds the source of the listener from the Gateway resource.
+	Source v1alpha2.Listener
+	// Valid shows whether the listener is valid.
+	// FIXME(pleshakov) For now, only capture true/false without any error message.
+	Valid bool
+	// Routes holds the routes attached to the listener.
+	Routes map[types.NamespacedName]*route
+	// AcceptedHostnames is an intersection between the hostnames supported by the listener and the hostnames
+	// from the attached routes.
+	AcceptedHostnames map[string]struct{}
+}
+
+// route represents an HTTPRoute.
+type route struct {
+	// Source is the source resource of the route.
+	// FIXME(pleshakov)
+	// For now, we assume that the source is only HTTPRoute. Later we can support more types - TLSRoute, TCPRoute and UDPRoute.
+	Source *v1alpha2.HTTPRoute
+
+	// ValidSectionNameRefs includes the sectionNames from the parentRefs of the HTTPRoute that are valid -- i.e.
+	// the Gateway resource has a corresponding valid listener.
+	ValidSectionNameRefs map[string]struct{}
+	// ValidSectionNameRefs includes the sectionNames from the parentRefs of the HTTPRoute that are invalid.
+	InvalidSectionNameRefs map[string]struct{}
+}
+
+// graph is a graph-like representation of Gateway API resources.
+// It is assumed that the Gateway resource is a single resource.
+type graph struct {
+	// Listeners holds listeners keyed by their names in the Gateway resource.
+	Listeners map[string]*listener
+	// Routes holds route resources.
+	Routes map[types.NamespacedName]*route
+}
+
+// buildGraph builds a graph from a store assuming that the Gateway resource has the gwNsName namespace and name.
+func buildGraph(store *store, gwNsName types.NamespacedName) *graph {
+	listeners := buildListeners(store.gw)
+	routes := buildRoutesAndBindToListeners(store.httpRoutes, gwNsName, listeners)
+
+	return &graph{
+		Listeners: listeners,
+		Routes:    routes,
+	}
+}
+
+func buildRoutesAndBindToListeners(
+	httpRoutes map[types.NamespacedName]*v1alpha2.HTTPRoute,
+	gwNsName types.NamespacedName,
+	listeners map[string]*listener,
+) map[types.NamespacedName]*route {
+	routes := make(map[types.NamespacedName]*route)
+
+	for _, ghr := range httpRoutes {
+		if len(ghr.Spec.ParentRefs) == 0 {
+			// ignore HTTPRoute without refs
```

Comment:
We cannot set its status because it doesn't reference our Gateway.

I think we log something to our log, but not sure if it will be useful - logs are only visible to cluster operator, not application developers (the ones who create routes). 

 I think it is clear from the API that each HTTPRoute must have a parentRef, so I don't expect this case to be common.